cirq/experiments/fidelity_estimation.py

```diff
@@ -372,3 +394,225 @@ def least_squares_xeb_fidelity_from_probabilities(
+        raise ValueError(
+            "`circuits` should be a sequence of circuits each operating on the same two qubits."
+        )
+    q0, q1 = all_all_qubits
+
+    tasks = []
+    for cycle_depth in cycle_depths:
+        for circuit_i, circuit in enumerate(circuits):
+            circuit_depth = cycle_depth * 2 + 1
+            assert circuit_depth <= len(circuit)
+            tcircuit = circuit[:circuit_depth]
+            pcircuit = tcircuit + ops.measure(q0, q1)
+            tasks.append(
+                _Sample2qXEBTask(
+                    cycle_depth=cycle_depth, circuit_i=circuit_i, prepared_circuit=pcircuit
+                )
+            )
+
+    n_tasks = len(tasks)
+    b_tasks = [tasks[i : i + batch_size] for i in range(0, n_tasks, batch_size)]
+
+    run_batch = _SampleInBatches(sampler=sampler, repetitions=repetitions)
+    with ThreadPoolExecutor(max_workers=2) as pool:
+        futures = [pool.submit(run_batch, task) for task in b_tasks]
+
+        records = []
+        with tqdm.tqdm(total=n_tasks) as progress:
+            for future in concurrent.futures.as_completed(futures):
+                records += future.result()
+                progress.update(batch_size)
+
+    return pd.DataFrame(records).set_index(['circuit_i', 'cycle_depth'])
+
+
+def _simulate_2q_xeb_circuit(task: Dict[str, Any]):
+    """Helper function for simulating a given (circuit, cycle_depth)."""
+    circuit_i = task['circuit_i']
+    cycle_depth = task['cycle_depth']
+    circuit = task['circuit']
+    param_resolver = task['param_resolver']
+
+    circuit_depth = cycle_depth * 2 + 1
+    assert circuit_depth <= len(circuit)
+    tcircuit = circuit[:circuit_depth]
+    tcircuit = protocols.resolve_parameters_once(tcircuit, param_resolver=param_resolver)
+
+    pure_sim = sim.Simulator()
+    psi = cast(sim.StateVectorTrialResult, pure_sim.simulate(tcircuit))
+    psi = psi.final_state_vector
+    pure_probs = np.abs(psi) ** 2
+
+    return {
+        'circuit_i': circuit_i,
+        'cycle_depth': cycle_depth,
+        'pure_probs': pure_probs,
+    }
+
+
+def simulate_2q_xeb_circuits(
+    circuits: Sequence['cirq.Circuit'],
+    cycle_depths: Sequence[int],
+    param_resolver: 'cirq.ParamResolverOrSimilarType' = None,
+    pool: Optional['multiprocessing.pool.Pool'] = None,
+):
+    """Simulate two-qubit XEB circuits.
+
+    These ideal probabilities can be benchmarked against potentially noisy
+    results from `sample_2q_xeb_circuits`.
+
+    Args:
+        circuits: A library of two-qubit circuits generated from
+            `random_rotations_between_two_qubit_circuit` of sufficient length
+            for `cycle_depths`.
+        cycle_depths: A sequence of cycle depths at which we will truncate each
+            of the `circuits` to simulate.
+        param_resolver: If circuits contain parameters, resolve according
+            to this ParamResolver prior to simulation
+        pool: If provided, execute the simulations in parallel.
```

Comment:
Is this formatting good? Looks narrow.

Reply:
I thought we stuck to 80 character width for docstrings still. Or are new docstrings supposed to be wide? @balopat

Reply:
docstrings can be up to 100 too

Reply:
re-wrapped